File: cpp/src/prime_sieve.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <iomanip>

const size_t LIMIT = 100'000'000;

std::vector<size_t> sieve_of_eratosthenes(size_t limit) {
    std::vector<bool> is_prime(limit + 1, true);
    is_prime[0] = false;
    is_prime[1] = false;
    
    size_t sqrt_limit = static_cast<size_t>(std::sqrt(limit));
    
    for (size_t i = 2; i <= sqrt_limit; i++) {
        if (is_prime[i]) {
            for (size_t j = i * i; j <= limit; j += i) {
                is_prime[j] = false;
            }
        }
    }
    
    std::vector<size_t> primes;
    for (size_t i = 0; i <= limit; i++) {
        if (is_prime[i]) {
            primes.push_back(i);
        }
    }
    
    return primes;
}

int main() {
    // Warm-up with smaller limit
    auto warmup = sieve_of_eratosthenes(1'000'000);
    
    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    auto primes = sieve_of_eratosthenes(LIMIT);
    auto end = std::chrono::high_resolution_clock::now();
    
    std::chrono::duration<double> duration = end - start;
    
    std::cout << std::fixed << std::setprecision(6) << duration.count() << std::endl;
    std::cerr << "Number of primes: " << primes.size() << std::endl;
    
    return 0;
}

File: rust/src/prime_sieve.rs
use std::time::Instant;

const LIMIT: usize = 100_000_000;

fn sieve_of_eratosthenes(limit: usize) -> Vec<usize> {
    let mut is_prime = vec![true; limit + 1];
    is_prime[0] = false;
    is_prime[1] = false;
    
    let sqrt_limit = (limit as f64).sqrt() as usize;
    
    for i in 2..=sqrt_limit {
        if is_prime[i] {
            let mut j = i * i;
            while j <= limit {
                is_prime[j] = false;
                j += i;
            }
        }
    }
    
    is_prime.iter()
        .enumerate()
        .filter(|(_, &is_p)| is_p)
        .map(|(i, _)| i)
        .collect()
}

fn main() {
    // Warm-up with smaller limit
    let _ = sieve_of_eratosthenes(1_000_000);
    
    // Benchmark
    let start = Instant::now();
    let primes = sieve_of_eratosthenes(LIMIT);
    let duration = start.elapsed();
    
    println!("{:.6}", duration.as_secs_f64());
    eprintln!("Number of primes: {}", primes.len());
}

